local-volume/provisioner/pkg/deleter/blockcleaner.go

```diff
@@ -0,0 +1,247 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deleter
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/external-storage/local-volume/provisioner/pkg/common"
+	"k8s.io/api/core/v1"
+	"os/exec"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	// ZeroDev is refers to zero device.
+	ZeroDev = "/dev/zero"
+	// DefaultBlockSize is the default block size to use in dd
+	DefaultBlockSize = "4096"
+	// ShredCommand the shred cli.
+	ShredCommand = "shred"
+	// MkfsCommand is the mkfs executable.
+	MkfsCommand = "mkfs"
+	// DDCommand is the dd executable.
+	DDCommand = "dd"
+	// WipeComand is the dd executable.
+	WipeComand = "wipefs"
+	// ShellCommand is the command to invoke shell
+	ShellCommand = "sh"
+
+	// DDFullErrExpected Error ouptput expected when running dd to fill a device
+	DDFullErrExpected = "No space left on device"
+	// NoErrExpected No error expected from running command
+	NoErrExpected = ""
+)
+
+// BlockCleaner Interface for invoking async script to do cleanup of block device.
+type BlockCleaner interface {
+	// DoCleanup deletes
+	DeleteBlockPV(pv *v1.PersistentVolume, mountPath string, config common.MountConfig) error
+	isRunning(pvName string) bool
+	markRunning(pvName string)
+	markDone(pvName string)
+}
+
+var _ BlockCleaner = &BlockCleanerImpl{}
+
+// BlockCleanerImpl Implementation of BLockCleaner interface
+type BlockCleanerImpl struct {
+	*common.RuntimeConfig
+	mutex     sync.Mutex
+	procTable map[string]time.Time
+}
+
+// NewBlockCleaner returns a BlockCleaner
+func NewBlockCleaner(config *common.RuntimeConfig) *BlockCleanerImpl {
+	return &BlockCleanerImpl{RuntimeConfig: config,
+		procTable: make(map[string]time.Time),
+	}
+}
+
+func (v *BlockCleanerImpl) isRunning(pvName string) bool {
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+	_, ok := v.procTable[pvName]
+	return ok
+}
+
+func (v *BlockCleanerImpl) markRunning(pvName string) {
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+	v.procTable[pvName] = time.Now()
+}
+
+func (v *BlockCleanerImpl) markDone(pvName string) {
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+	delete(v.procTable, pvName)
+}
+
+// DeleteBlockPV is meant to execute cleanup concurrently with other such cleanup processes.
+func (v *BlockCleanerImpl) DeleteBlockPV(pv *v1.PersistentVolume, mountPath string, config common.MountConfig) error {
+	iters, expErr, exe, args, err := v.getBlockDeleteCommand(mountPath, config)
+	if err != nil {
+		return err
+	}
+
+	if v.isRunning(pv.Name) {
+		// Run in progress, nothing to do,
+		return nil
+	}
+	v.markRunning(pv.Name)
```

Comment:
It may be good to also log an event on the PV when we start the cleanup and warn that it could take a long time.  Otherwise users could get impatient and just delete the PV object.  Actually, what happens if the user deletes the PV object and we're still cleaning the device?  Will the discovery kick in and recreate the PV?  Maybe we need some sort of check there.  I also still wonder if we need to persist the status somewhere and/or prevent the PV from being destroyed by the user unless some force flag is set.

Reply:
OK, will log. I will also put in a check before creating a PV to see if its currently being cleaned. If it is, then a log message will be generated and the creation blocked. I feel this largely covers reasonable use cases for now.

Having something in the API server/API itself for Local PVs to prevent deletion might be the longterm solution here. This is much like pods are not terminated until they get kubelet to shutdown the docker/rkt containers. Perhaps the PV needs a CLEANING and CLEANED phase (not just RELEASED) and refuse any un-forced deletion in CLEANING phase.